File: src/main/dev-bundle/config/templates/src/around-drop-target-overlay.ts
import { css, html, LitElement } from 'lit';

class AroundDropTargetOverlay extends LitElement {
    static get styles() {
        return css`
          :host {
            --overlay-visibility: hidden;
            
            position: absolute;
            top: 0;
            left: 0;
            right: 0;
            bottom: 0;
            width: 100%;
            height: 100%;
            z-index: 999999;
            visibility: var(--overlay-visibility, hidden);
          }

          #around-overlay {
            --top-left: calc(var(--around-overlay-min-dim) * 1px / 2);
            --bottom-right: calc(100% - var(--top-left));
            
            position: relative;
            left: 0;
            top: 0;
            right: 0;
            bottom: 0;
            width: 100%;
            height: 100%;
          }

          .drop-zone {
            box-sizing: border-box;
            position: absolute;
            opacity: 0.2;
          }

          #before-zone {
            left: 0;
            top: 0;
            width: 50%;
            height: 100%;
            clip-path: polygon(0 0, var(--top-left) var(--top-left), var(--top-left) var(--bottom-right), 0 100%);
            border-left: 5px solid lightgray;
            background-color: green;
          }

          #above-zone {
            left: 0;
            top: 0;
            width: 100%;
            height: 100%;
            clip-path: polygon(0 0, var(--top-left) var(--top-left), var(--bottom-right) var(--top-left), 100% 0);
            border-top: 5px solid lightgray;
            background-color: purple;
          }

          #below-zone {
            left: 0;
            bottom: 0;
            width: 100%;
            height: 50%;
            clip-path: polygon(0 100%, var(--top-left) var(--bottom-right), var(--bottom-right) var(--bottom-right), 100% 100%);
            border-bottom: 5px solid lightgray;
            background-color: orange;
          }

          #after-zone {
            right: 0;
            top: 0;
            width: 50%;
            height: 100%;
            clip-path: polygon(100% 0, var(--bottom-right) var(--top-left), var(--bottom-right) var(--bottom-right), 100% 100%);
            border-right: 5px solid lightgray;
            background-color: red;
          }

          #before-zone.v-drag-over-target,
          #above-zone.v-drag-over-target,
          #below-zone.v-drag-over-target,
          #after-zone.v-drag-over-target {
            opacity: 0.9;
          }
          
          #before-zone.v-drag-over-target {
            border-left: 3px solid rgba(127, 127, 255, 0.3);
          }

          #above-zone.v-drag-over-target {
            border-top: 3px solid rgba(127, 127, 255, 0.3);
          }

          #below-zone.v-drag-over-target {
            border-bottom: 3px solid rgba(127, 127, 255, 0.3);
          }

          #after-zone.v-drag-over-target {
            border-right: 3px solid rgba(127, 127, 255, 0.3);
          }
        `;
    }

    constructor() {
        super();

        console.log('HERE I WAS')
        // new ResizeObserver(entries=> {
        //     for (const entry of entries) {
        //         const { width, height } = entry.contentRect;
        //         console.log('HERE I AM')
        //         entry.target.style.setProperty('--around-overlay-min-dim', Math.min(width, height));
        //     }
        // }).observe(window.document.getElementById("around-overlay"));
    }

    render() {
        return html`
            <div id="around-overlay">
                <div id="before-zone" class="drop-zone"></div>
                <div id="above-zone" class="drop-zone"></div>
                <div id="below-zone" class="drop-zone"></div>
                <div id="after-zone" class="drop-zone"></div>
            </div>
        `;
    }
}

customElements.define('around-drop-target-overlay', AroundDropTargetOverlay);

File: src/main/java/com/vaadin/flux/ui/dragdrop/overlay/AroundDropTargetOverlay.java
package com.vaadin.flux.ui.dragdrop.overlay;

import com.vaadin.flow.component.AttachEvent;
import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flux.ui.dragdrop.AbstractDropTargetOverlay;
import com.vaadin.flux.ui.dragdrop.DragDropKit;
import com.vaadin.flux.ui.dragdrop.DropZone;

import java.util.function.Consumer;

/**
 * <pre>
 * wide                     tall
 * +--------------------+   +-----------+
 * |\                  /|   |\         /|
 * | \                / |   | \       / |
 * |  \              /  |   |  \     /  |
 * |   \            /   |   |   \   /   |
 * |    \__________/    |   |    \ /    |
 * |    /          \    |   |     |     |
 * |   /            \   |   |     |     |
 * |  /              \  |   |     |     |
 * | /                \ |   |     |     |
 * |/                  \|   |     |     |
 * +--------------------+   |    / \    |
 *                          |   /   \   |
 *                          |  /     \  |
 *                          | /       \ |
 *                          |/         \|
 *                          +-----------+
 * </pre>
 */
public class AroundDropTargetOverlay extends AbstractDropTargetOverlay {
    public static final String OVERLAY_CLASS_NAME = "around-drop-target-overlay";
    public static final String DROP_ZONE_CLASS_NAME = "around-drop-zone";

    /**
     * Create an {@code AroundDropTargetOverlay} for a component.
     *
     * @param overlaidComponent target component overlaid during drag operation
     * @param beforeConsumer method to handle drag source being dropped on the "before" drop zone
     * @param aboveConsumer method to handle drag source being dropped on the "above" drop zone
     * @param belowConsumer method to handle drag source being dropped on the "below" drop zone
     * @param afterConsumer method to handle drag source being dropped on the "after" drop zone
     */
    private AroundDropTargetOverlay(Component overlaidComponent,
                                    Consumer<Component> beforeConsumer,
                                    Consumer<Component> aboveConsumer,
                                    Consumer<Component> belowConsumer,
                                    Consumer<Component> afterConsumer) {
        super(overlaidComponent, OVERLAY_CLASS_NAME);

        var beforeDiv = new Div();
        beforeDiv.addClassNames(DROP_ZONE_CLASS_NAME, "before-zone");

        var aboveDiv = new Div();
        aboveDiv.addClassNames(DROP_ZONE_CLASS_NAME, "above-zone");

        var belowDiv = new Div();
        belowDiv.addClassNames(DROP_ZONE_CLASS_NAME, "below-zone");

        var afterDiv = new Div();
        afterDiv.addClassNames(DROP_ZONE_CLASS_NAME, "after-zone");

        var content = getContent();
        content.setSizeFull();
        content.add(beforeDiv);
        content.add(aboveDiv);
        content.add(belowDiv);
        content.add(afterDiv);

        addDropZone(new DropZone(beforeDiv, beforeConsumer));
        addDropZone(new DropZone(aboveDiv, aboveConsumer));
        addDropZone(new DropZone(belowDiv, belowConsumer));
        addDropZone(new DropZone(afterDiv, afterConsumer));

        addAttachListener(this::onAroundDropTargetOverlayAttach);
    }

    private void onAroundDropTargetOverlayAttach(AttachEvent attachEvent) {
        attachEvent.getUI().getPage().executeJs("""
                  $0.style.setProperty('--around-overlay-min-dim', '100');
                  new ResizeObserver(entries => {
                    for (const entry of entries) {
                      const { width, height } = entry.contentRect;
                      console.log('>>>>> ' + $0.parentElement.tagName + ': ' + width + 'x' + height);
                      entry.target.style.setProperty('--around-overlay-min-dim', Math.min(width, height));
                    }
                  }).observe($0);
                """, this.getElement());
        // TODO call the above when attaching, not just when resizing?
    }

    /**
     * Factory method to create an {@code AroundDropTargetOverlay} for a component.
     *
     * @param overlaidComponent target component overlaid during drag operation
     * @param dragSourceClass class of drag source component to be received by drop target overlay
     * @param dropTargetOverlayVisibilityClassName name of CSS class used to control the drop target overlay's visibility
     * @param beforeConsumer method to handle drag source being dropped on the "before" drop zone
     * @param aboveConsumer method to handle drag source being dropped on the "above" drop zone
     * @param belowConsumer method to handle drag source being dropped on the "below" drop zone
     * @param afterConsumer method to handle drag source being dropped on the "after" drop zone
     * @param <S> Type of component being dragged
     */
    @SuppressWarnings("java:S107")
    public static <S extends Component>
    void initForMoveDrop(Component overlaidComponent, Class<S> dragSourceClass,
                         String dropTargetOverlayVisibilityClassName,
                         Consumer<Component> beforeConsumer,
                         Consumer<Component> aboveConsumer,
                         Consumer<Component> belowConsumer,
                         Consumer<Component> afterConsumer) {
        var dropTargetOverlay = new AroundDropTargetOverlay(overlaidComponent,
                beforeConsumer, aboveConsumer, belowConsumer, afterConsumer);
        dropTargetOverlay.addClassNames(dropTargetOverlayVisibilityClassName);

        overlaidComponent.getStyle().set("position", "relative");
        overlaidComponent.getElement().appendChild(dropTargetOverlay.getElement());

        DragDropKit.initForMoveDrop(dropTargetOverlay, dragSourceClass);
    }
}
